### PR TITLE
Three work items to improve autoenrollment

### PR DESCRIPTION
**Pull request by mehalshah**
- When new teachers get created, see if we have to associate them with any existing enrollments
- When the user enrolls but is not logged in, we'll still associate the enrollment with the user if the email associated with a teacher
- When the user enrolls but is not logged in, but later does attendance with the email, we'll update the user field

### CODE REVIEW

File: dashboard/app/models/pd/enrollment.rb [hunk 1]
```diff
@@ -59,6 +59,7 @@ class Pd::Enrollment < ActiveRecord::Base
 
   before_validation :autoupdate_user_field
   after_create :enroll_in_corresponding_online_learning
+  after_save :enroll_in_corresponding_online_learning, if: -> {user_id_changed? || email_changed?}
```
Comment (aoby) [inline dashboard/app/models/pd/enrollment.rb]: remove the `after_create` on line 61, as it's now covered in `after_save`

File: dashboard/test/models/pd/enrollment_test.rb [hunk 2]
```diff
@@ -339,6 +339,32 @@ class Pd::EnrollmentTest < ActiveSupport::TestCase
     end
   end
 
+  test 'enrolling in class, and then later having the user field updated enrolls in online learning' do
+    teacher = create :teacher
+    create :plc_course, name: 'CSP Support'
+    workshop = create :pd_workshop, course: Pd::Workshop::COURSE_CSP
+    enrollment = create :pd_enrollment, user: nil, workshop: workshop
+    enrollment.update(user: teacher)
+    assert_equal 'CSP Support', Plc::UserCourseEnrollment.find_by(user: teacher).plc_course.name
```
Comment (aoby) [inline dashboard/test/models/pd/enrollment_test.rb]: how about:

```ruby
user_course_enrollment = assert_creates Plc::UserCourseEnrollment do
  enrollment.update!(user: teacher)
end
assert_equal 'CSP Support', user_course_enrollment.plc_course.name
```

Same below

File: dashboard/test/models/pd/enrollment_test.rb [hunk 3]
```diff
@@ -339,6 +339,32 @@ class Pd::EnrollmentTest < ActiveSupport::TestCase
...
+
+  test 'enrolling in class while not logged in still associates the user' do
+    teacher = create :teacher
+    create :plc_course, name: 'CSP Support'
+    workshop = create :pd_workshop, course: Pd::Workshop::COURSE_CSP
+    create :pd_enrollment, user: nil, workshop: workshop, email: teacher.email
+    assert_equal 'CSP Support', Plc::UserCourseEnrollment.find_by(user: teacher).plc_course.name
+  end
+
+  test 'enrolling in class without an account creates an account when the user is created' do
```
Comment (aoby) [inline dashboard/test/models/pd/enrollment_test.rb]: creates a PLC Enrollment, not an account, right?

File: dashboard/app/models/pd/enrollment.rb [hunk 4]
```diff
@@ -226,6 +227,8 @@ def autoupdate_user_field
   end
 
   def enroll_in_corresponding_online_learning
+    autoupdate_user_field if email_changed?
```
Comment (aoby) [inline dashboard/app/models/pd/enrollment.rb]: Note, this is already called [before_validation](https://github.com/code-dot-org/code-dot-org/blob/0d5d2c65a4181b9ab60f14e9038671cef9c51254/dashboard/app/models/pd/enrollment.rb#L60) in all cases, which is called before save as long as validation is not suppressed, so this line shouldn't be needed.  

I think we're good just by switching from `after_create` to `after_save` (with constraints) above. Try running the tests without this.

**Comment by ryansloan** [comment]
🙌

**Pull request merged**